File: mobile/lib/models/quest.dart
import 'user_rating.dart';

/// Quests: a system entirely separate from achievements/rating's own
/// models -- mirrors the backend's AvailableQuestOut/QuestRoundOut/
/// QuestAnswerOut exactly. The backend decides eligibility, round content,
/// and the reward; this screen only renders what it's given and reports
/// {word_id, is_correct}, same principle every exercise already follows.
class AvailableQuest {
  final int id;
  final String name;
  final String wordLevelName;
  final String exerciseKey;
  final int rewardPoints;
  final bool available;

  /// How many successful attempts count as "done for today" (the admin's
  /// own per-quest setting), how many the user has actually made since the
  /// daily reset, and whether that target is met. All three come straight
  /// from the backend -- the client never counts completions itself.
  final int dailyTarget;
  final int completedToday;
  final bool isDoneToday;

  AvailableQuest({
    required this.id,
    required this.name,
    required this.wordLevelName,
    required this.exerciseKey,
    required this.rewardPoints,
    required this.available,
    required this.dailyTarget,
    required this.completedToday,
    required this.isDoneToday,
  });

  factory AvailableQuest.fromJson(Map<String, dynamic> json) {
    return AvailableQuest(
      id: json['id'] as int,
      name: json['name'] as String,
      wordLevelName: json['word_level_name'] as String,
      exerciseKey: json['exercise_key'] as String,
      rewardPoints: json['reward_points'] as int,
      available: json['available'] as bool,
      dailyTarget: json['daily_target'] as int,
      completedToday: json['completed_today'] as int,
      isDoneToday: json['is_done_today'] as bool,
    );
  }
}

/// The season a user is currently playing, as the quests section shows it.
/// Mirrors the backend's SeasonOut: `iconUrl` is the picture an admin
/// uploaded in the season settings, and is the ONLY season image the app
/// ever renders.
class QuestSeason {
  final int id;
  final String name;
  final String? iconUrl;
  final DateTime startsAt;
  final DateTime? endsAt;

  QuestSeason({
    required this.id,
    required this.name,
    required this.iconUrl,
    required this.startsAt,
    required this.endsAt,
  });

  factory QuestSeason.fromJson(Map<String, dynamic> json) {
    return QuestSeason(
      id: json['id'] as int,
      name: json['name'] as String,
      iconUrl: json['icon_url'] as String?,
      startsAt: DateTime.parse(json['starts_at'] as String).toLocal(),
      endsAt: json['ends_at'] == null ? null : DateTime.parse(json['ends_at'] as String).toLocal(),
    );
  }
}

/// Everything the "Квесты сезона" block on Главная and the season quests
/// screen render, in one round trip -- mirrors the backend's
/// SeasonQuestOverviewOut.
///
/// Every number here is the backend's own: the season comes from the
/// season system, the rank position is the same one the Profile screen
/// shows, the points are counted from the rating ledgers, and quest
/// progress is counted from the quest completions themselves. This screen
/// computes none of it and holds no second copy of it.
class SeasonQuestOverview {
  final QuestSeason? season;

  /// The season's own timeline, both already rounded by the backend:
  /// whole days left, and how many whole days the season runs in total.
  /// Both null when the season has no scheduled end -- then there is
  /// neither a countdown nor a range to draw. Deliberately separate from
  /// quest progress below: the two mean different things and must never
  /// share a bar.
  final int? daysLeft;
  final int? daysTotal;

  /// Rating points earned today: newly learned words plus quest rewards.
  final int pointsToday;
  final int totalPoints;
  final RankSummary? rank;

  /// The user's real place within their own rank.
  final int? rankPosition;

  final int questsDoneToday;
  final int questsTotal;

  /// The permanent "изучение новых слов" quest: what one newly learned
  /// word is worth right now, and how many were learned today.
  final int pointsPerLearnedWord;
  final int wordsLearnedToday;

  final List<AvailableQuest> quests;

  SeasonQuestOverview({
    required this.season,
    required this.daysLeft,
    required this.daysTotal,
    required this.pointsToday,
    required this.totalPoints,
    required this.rank,
    required this.rankPosition,
    required this.questsDoneToday,
    required this.questsTotal,
    required this.pointsPerLearnedWord,
    required this.wordsLearnedToday,
    required this.quests,
  });

  factory SeasonQuestOverview.fromJson(Map<String, dynamic> json) {
    return SeasonQuestOverview(
      season: json['season'] == null ? null : QuestSeason.fromJson(json['season'] as Map<String, dynamic>),
      daysLeft: json['days_left'] as int?,
      daysTotal: json['days_total'] as int?,
      pointsToday: json['points_today'] as int,
      totalPoints: json['total_points'] as int,
      rank: json['rank'] == null ? null : RankSummary.fromJson(json['rank'] as Map<String, dynamic>),
      rankPosition: json['rank_position'] as int?,
      questsDoneToday: json['quests_done_today'] as int,
      questsTotal: json['quests_total'] as int,
      pointsPerLearnedWord: json['points_per_learned_word'] as int,
      wordsLearnedToday: json['words_learned_today'] as int,
      quests: (json['quests'] as List<dynamic>)
          .map((e) => AvailableQuest.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}

/// `payload` is the SAME round shape Lessons already use for this
/// exercise_key (ExerciseWordsOut / TrueOrFalseRoundOut / BuildWordRoundOut
/// / SpeakingWordRoundOut / ListenWordRoundOut) -- QuestAttemptScreen parses
/// it with the exact same model classes the Lesson screens already use.
class QuestRound {
  final int questId;
  final int wordId;
  final String exerciseKey;
  final Map<String, dynamic> payload;

  QuestRound({required this.questId, required this.wordId, required this.exerciseKey, required this.payload});

  factory QuestRound.fromJson(Map<String, dynamic> json) {
    return QuestRound(
      questId: json['quest_id'] as int,
      wordId: json['word_id'] as int,
      exerciseKey: json['exercise_key'] as String,
      payload: json['payload'] as Map<String, dynamic>,
    );
  }
}

class QuestAnswerResult {
  final int wordId;
  final int score;
  final bool isCorrect;
  final int rewardGranted;

  QuestAnswerResult({required this.wordId, required this.score, required this.isCorrect, required this.rewardGranted});

  factory QuestAnswerResult.fromJson(Map<String, dynamic> json) {
    return QuestAnswerResult(
      wordId: json['word_id'] as int,
      score: json['score'] as int,
      isCorrect: json['is_correct'] as bool,
      rewardGranted: json['reward_granted'] as int,
    );
  }
}

File: backend/app/schemas/quest.py
from typing import Any

from pydantic import BaseModel

from app.schemas.rating import RankPublicOut, SeasonOut


class QuestOut(BaseModel):
    """Admin Web's view of one Quest definition."""

    id: int
    name: str
    word_level_id: int
    word_level_name: str
    exercise_key: str
    reward_points: int
    daily_target: int
    enabled: bool
    order: int


class ReorderQuestsIn(BaseModel):
    quest_ids: list[int]


class AvailableQuestOut(BaseModel):
    """One quest as shown to a user -- the definition plus whether they
    currently have an eligible, not-yet-used-today word for it in the
    given dictionary. `available=False` quests are still listed (so the
    user can see what exists), just not attemptable right now.

    `completed_today` counts this quest's own UserQuestWordDay rows for
    today, i.e. how many words it has successfully consumed since the
    daily reset; `is_done_today` is simply that count reaching
    `daily_target`. Neither is stored anywhere -- both are counted from
    the rows the quest flow already writes."""

    id: int
    name: str
    word_level_name: str
    exercise_key: str
    reward_points: int
    available: bool
    daily_target: int
    completed_today: int
    is_done_today: bool


class QuestRoundOut(BaseModel):
    """A round for exactly one target word. `payload` is the SAME Out
    schema Lessons already use for this exercise_key (ExerciseWordsOut /
    TrueOrFalseRoundOut / BuildWordRoundOut / SpeakingWordRoundOut /
    ListenWordRoundOut, dumped to a plain dict) -- the client already
    knows how to parse each shape from the Lessons flow and picks the
    right parser by `exercise_key` here, exactly like it already picks
    the right screen by exercise_key for Lessons."""

    quest_id: int
    word_id: int
    exercise_key: str
    payload: dict[str, Any]


class QuestAnswerIn(BaseModel):
    word_id: int
    is_correct: bool


class SeasonQuestOverviewOut(BaseModel):
    """Everything the home screen's "Квесты сезона" block and the season
    quests screen render, in one round trip.

    A read-only VIEW, not a system of its own: every number below is
    counted from tables that already exist -- the active Season
    (app/rating/), UserRating/Rank (the same rank_position the Profile
    screen shows), UserWordPoints, UserQuestWordDay and Quest. Nothing
    here grants points, writes progress, or defines a second notion of
    "completed".
    """

    # The active season, or null when no season is running right now.
    # `days_left` and `days_total` are both null for a season with no
    # scheduled end (it ends when an admin says so, so there is neither a
    # countdown nor a range to show). Together they are the season's own
    # timeline: how much of it is gone and how much is left -- deliberately
    # NOT the same thing as quest progress, which is counted below.
    season: SeasonOut | None
    days_left: int | None
    days_total: int | None

    # Rating points this user earned TODAY: the words they learned plus
    # the quest rewards they collected, on the same Asia/Dushanbe day
    # boundary quests themselves reset on.
    points_today: int
    total_points: int
    rank: RankPublicOut | None
    # The user's real 1-based place within their own rank -- the very same
    # number GET /users/me/rating reports, never a second calculation.
    rank_position: int | None

    # How many of the enabled quests have reached their daily target
    # today, out of how many exist.
    quests_done_today: int
    quests_total: int

    # The permanent "изучение новых слов" quest: the reward each newly
    # learned word grants right now, and how many the user has learned
    # today. It has no attempt flow of its own -- the reward fires
    # wherever a word is first learned.
    points_per_learned_word: int
    words_learned_today: int

    # Every enabled quest, exactly as GET /quests lists them.
    quests: list[AvailableQuestOut]


class QuestAnswerOut(BaseModel):
    word_id: int
    score: int
    is_correct: bool
    # 0 when the answer was wrong, or when the word had already been used
    # in some other quest today (a race) -- never negative, never implies
    # a second grant happened.
    reward_granted: int
